Adding server processing

mayonezzzzzzz opened the pull request:
Проект был доработан в соответствии с описанием задания. Были внесены все поправки, описанные в isuues и в предыдущем PR.

Последовательность изменений можно просмотреть в истории коммитов.

mayonezzzzzzz committed:
feat(client): specifying the path to find images and save messages

User can write a path to "path.txt" file.

If the file is empty, default path to the current directory is using.

mayonezzzzzzz committed:
fix(server): fixed possible 0 retutn in std::thread::hardware_concurrency() function.

mayonezzzzzzz committed:
feat(client): added ability to send text for overlay

Every message contains not only image, but also a text for overlay

Text is sending with the first (and maybe the only) part of the image

mayonezzzzzzz committed:
fix(client): fixed body limit problem

Separator is also included in calculation of part size

mayonezzzzzzz committed:
fix(client): avoiding server resource consumption

Putting repeating method calls into a separate reference

mayonezzzzzzz committed:
fix(server): converting INT into UNSIGNED INT

mayonezzzzzzz committed:
fix(compression): passing constant values to the function

mayonezzzzzzz committed:
fix(client): fixed incorrect inserting values into vector

mayonezzzzzzz committed:
feat(server): added text overlay on the image

mayonezzzzzzz committed:
feat(server): added ability to receive text and image

fixed some moments: passing constant values to the function, using std::move() instead of copying

BREACKING CHANGE: text for overlay is receiving in the first request, then receiving only image parts

mayonezzzzzzz committed:
fix: changes in CmakeLists.txt files

Removed optional check if(Boost_FOUND) on client and server

Creating only 1 project in main CmakeLists.txt

Removed finding Boost packages in client and server CmakeLists.txt

Finding OpenCV library

mayonezzzzzzz committed:
fix: right calculation of the text position while overlaying it

mayonezzzzzzz committed:
feat(server)!: limiting the number of requests

Limiting the number of simultneously processed requests with responses when limit is reached

mayonezzzzzzz committed:
Merge branch 'adding_server_processing' of https://github.com/mayonezzzzzzz/Client-Server-Application into adding_server_processing

mayonezzzzzzz committed:
feat(client): added parametrization

Parameters:
 -help
 -address
 -port
 -images_path
 -text
 -save_path

mayonezzzzzzz committed:
feat(CmakeLists.txt): added program_options component

mayonezzzzzzz committed:
feat(server): added parametrization

Parameters:
 -help
 -port
 -max_requests

Also created std::atomic to track available requests for processing

mayonezzzzzzz committed:
feat(CmakeLists.txt): added new component to server resources

mayonezzzzzzz committed:
fix(logging): fixed logging into the file

mayonezzzzzzz committed:
fix(client): added "endline" character after message

mayonezzzzzzz committed:
feat: аdding abbreviations to parameters names

mayonezzzzzzz committed:
fix(parameters): fixed an error whith images_path

Fixed an error where a path with spaces (" ") couldn't be readed completely

mayonezzzzzzz committed:
fix(CmakeLists.txt): removing absolute path to the boost library

using CMAKE_PREFIX_PATH to specify the path where to find Boost library

In my case: -DCMAKE_PREFIX_PATH="C:/Program Files/Boost/boost_1_86_0/boost_1_86_0"

mayonezzzzzzz committed:
fix(Readme): changed Readme file

Less information about developing, more about project and building it

mayonezzzzzzz committed:
docs: added parameters description

mayonezzzzzzz committed:
feat(address): changed address on server to accept connections from all interfaces

antekiv commented on CMakeLists.txt:
```diff
@@ -4,8 +4,13 @@ project(Client_Server_Application)
+message(STATUS "Boost library dirs: ${Boost_LIBRARY_DIRS}")
+
+include_directories(${Boost_INCLUDE_DIRS})
+link_directories(${Boost_LIBRARY_DIRS})
```
Это подключает зависимости во все проекты. Например, если в новом проекте не будет буста, то он все равно подключится.

Для того, чтобы это лучше контролировать, лучше всего непосредственно в самом проекте подключать через `target_*` аналоги методов зависимости

antekiv commented on client/src/HTTP_Client.cpp:
```diff
@@ -1,18 +1,57 @@
+    if (!params.images_path.empty()) {
+        if (!checkDirectory(params.images_path)) {
+            std::cerr << "Using default image directory\n";
+            params.images_path = ".";
```
как вариант в опциях по умолчанию можно было выставить значение

antekiv commented on client/src/Params.cpp:
```diff
@@ -0,0 +1,31 @@
+        ("address, a", po::value<std::string>(&params.address)->default_value("127.0.0.1"), "Server address")
+        ("port, p", po::value<std::string>(&params.port)->default_value("8080"), "Server port")
+        ("images-path, i", po::value<std::string>(&params.images_path)->default_value("C:\\Users\\Public\\Pictures"), "Path to the image file")
+        ("text, t", po::value<std::string>(&params.overlay_text)->default_value("default text"), "Text to overlay on the image")
```
такие пути не надо хардкодить... на другом компьютере он вряд-ли будет

antekiv commented on server/CMakeLists.txt:
```diff
@@ -1,23 +1,26 @@
 
+set(OpenCV_DIR "${CMAKE_SOURCE_DIR}/Libs/opencv/build")
+set(OpenCV_INCLUDE_DIR "${OpenCV_DIR}/include")
+set(OpenCV_LIBRARY "${OpenCV_DIR}/x64/vc16/lib/opencv_world4100.lib")
```
тут тоже хорошо бы по аналогии с бустом. В `openCV` должны быть модули, которые  знают про все эти зависимости и `find_package` должен его найти

antekiv commented on server/src/Compression.cpp:
```diff
@@ -38,7 +38,7 @@ bool decompress(const std::vector<char>& jpeg_data, std::vector<unsigned char>&
 
 // Функция для компрессии обработанного на сервере изображения
-bool compress(const std::vector<unsigned char>& image_data, std::vector<char>& jpeg_data, int width, int height, int& quality) {
+bool compress(const std::vector<unsigned char>& image_data, std::vector<char>& jpeg_data, const int width, const int height, const int quality) {
```
у тебя в одном месте используется вектор от `char`, тут от `unsigned char` - в чем разница?

mayonezzzzzzz replied:
Разницы нету, вектор из char переводится в unsigned char в теле функции decompress для передачи в функции libjpeg-а. 
Но, думаю, выделено это место не просто так, и мне более целесообразно сразу хранить все преданные изображения в vector из unsigned char.

antekiv commented:
более детально только вечером смогу посмотреть

antekiv commented on client/CMakeLists.txt:
```diff
@@ -1,14 +1,10 @@
 cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
```
Тут эта строка не нужна

antekiv commented on .gitignore:
```diff
@@ -1,4 +1,4 @@
-client/responses
 client/images
 Libs
+path_to_boost_for_building.txt
```
По-хорошему, почти все это должно быть в глобальном `.gitignore,` так как вся эта информация - твои отладочные директории/файлы, кроме чего-то генерируемого, например, `build`

как пример https://github.com/abseil/abseil-cpp/blob/master/.gitignore

antekiv commented on client/src/HTTP_Client.cpp:
```diff
@@ -1,18 +1,57 @@
+        std::cerr << "Error: Directory does not exist - " << path << "\n";
+        return false;
+    }
+    if (!std::filesystem::is_directory(path)) {
```
По-моему, достаточно одной проверки `is_directory`, так как она тоже делает проверку статуса файла (см. реализацию в стандартной библиотеке).

Кстати, существует 2 формы этих функций, которые кидают исключение - эти,
и, которая имеет вторым параметром `error_code` - в случае ошибки туда запишет ее.

В твоем случаем, если выбросится исключение, то программа упадет, так как оно не ловится.

antekiv commented on client/src/HTTP_Client.cpp:
```diff
@@ -1,18 +1,57 @@
+    return true;
+}
+
+void detectDirectories(ClientParams& params) {
```
Тут, как мне кажется, много копипаста, который делает одно и тоже. 
Я бы предложил вынести самое главное, и вызвать 2 раза от проверяемых путей (`std::filesystem::path`).

antekiv replied:
и это будет более прозрачно

antekiv commented on client/src/HTTP_Client.cpp:
```diff
@@ -1,18 +1,57 @@
+    std::filesystem::path images_path = params.images_path;
+    std::filesystem::path responses_path = std::filesystem::path(params.responses_path) / "responses";
 
     if (!std::filesystem::exists(images_path)) {
```
Эти проверки уже есть в `detectDirectories` . Надо немного продумать и 1 раз проверить и создать нужную директорию, если нет

antekiv commented on client/src/HTTP_Client.cpp:
```diff
@@ -22,21 +61,25 @@ int main() {
-            std::string const address = "127.0.0.1";
-            std::string const port = "8080";
+            //std::string const address = "127.0.0.1";
+            //std::string const port = "8080";
```
комментарии желательно удалять

a-zagaevskiy replied:
Мотивация тут следующая: в жизни такие куски кода постепенно устаревают и становятся неактуальными, да и вообще мало кто решается их раскомметировать. Полезней оставлять напоминалки: TODO, FIXIT. NOTE. Можно в комметариях писать примеры использования. В общем, все то, что может пригодится, но не сам код, который по неизвестной для членов твоей команды причине припрятали в закрома.

antekiv commented on client/src/HTTP_Client.cpp:
```diff
@@ -22,21 +61,25 @@ int main() {
             std::cout << "Available images:\n";
             for (const auto& image : std::filesystem::directory_iterator(images_path)) {
-                std::cout << image.path().filename() << "\n";
+                if (image.path().extension() == ".jpeg") {
```
Такой проверки достаточно. 
Для информации, можно еще так проверять - более надежно. 
[List of file signatures](https://en.wikipedia.org/wiki/List_of_file_signatures)

mayonezzzzzzz replied:
Про более надежный способ: в этом случае нужно было бы считать первые байты и сравнить с нужной сигнатурой/сигнатурами?

antekiv commented on client/src/HTTP_Client.cpp:
```diff
@@ -45,28 +88,64 @@ int main() {
             size_t offset = 0;
+            bool is_first_part = true;
+
             while (offset < image_size) {
```
На мой взгляд, очень замороченная и тяжело расширяемая передача. 
Если, например, изменится требование, и попросят передавать цвет текста, шрифт, размер шрифта, и т.д., то такое будет тяжело поддержать.

Для информации, я бы предложил почитать про https://ru.wikipedia.org/wiki/JSON

antekiv commented on server/src/HTTP_Server.cpp:
```diff
@@ -7,12 +9,18 @@
+
+        MAX_REQUESTS.store(params.max_requests);
+
+        const auto& numOfProcessors = std::max(1u, std::thread::hardware_concurrency());
```
Себя тут поправлю.  `std::thread::hardware_concurrency()` возвращает максимальное, количество потоков, которые могут нормально работать параллельно. Так как главный поток (main) является тоже потоком, то количество потоков, которые можно создать еще для оптимальной работы -  `std::thread::hardware_concurrency() - 1`

```cpp
const auto& numOfProcessors = std::max(1u, std::thread::hardware_concurrency() - 1);
```

antekiv commented on server/src/ImageProcessor.cpp:
```diff
@@ -0,0 +1,47 @@
+    std::cout << "Image dimensions: " << image.rows << "x" << image.cols << std::endl;
+
+    std::cout << "First bytes in image data: ";
+    for (size_t i = 0; i < std::min(image_data.size(), static_cast<size_t>(10)); ++i) {
```
тут мне кажется `std::min(image_data.size(), static_cast<size_t>(10))`, будет высчитываться на каждой итерации, так что ее лучше вынести в отдельную переменную

antekiv commented on server/src/Session.cpp:
```diff
@@ -47,46 +56,68 @@ void Session::sendNextPart(size_t offset, size_t total_size, std::string image_i
+            return;
+        }
+
+        active_requests++;
```
в данном случае лучше префиксные формы использовать, так как они внутри себя не создают временный объект (см.реализацию)

antekiv commented:
В целом, выглядит неплохо! Видно, что есть понимание в работе многих вещей, продолжай в том же духе! :)

a-zagaevskiy commented on server/src/Session.h:
```diff
@@ -9,6 +8,9 @@ namespace asio = boost::asio;
 using tcp = boost::asio::ip::tcp;
 
+// Максимальное количество обрабатываемых запросов
+extern std::atomic<size_t> MAX_REQUESTS;
```
Использовать atomic для статичного значения нецелесообразно. Это лишний механизм синхронизации, который будет мешаться потокам.

Заметь, что нет проблемы (undefined behaviour), когда ты из разных потоков обращаешься на чтение к переменной. Проблемы возникают в моделях "писатель"/"читатель" или когда время жизни переменных не синхронизированы со временем жизни потоков. Но это не этот случай.

a-zagaevskiy commented on server/src/MessageHandler.cpp:
```diff
@@ -1,11 +1,11 @@
 
-MessageHandler::MessageHandler(std::shared_ptr<asio::io_context> ioc) :
-    ioc(ioc), acceptor(*ioc, tcp::endpoint(asio::ip::make_address_v4("127.0.0.1"), 8080)) {
+MessageHandler::MessageHandler(std::shared_ptr<asio::io_context> ioc, const std::string port) :
```
Передача `port` по значению приводит к копированию строки, которая нигде не сохраняется, т.е. фактически выбрасывается. Лишние аллокации (особенно строк) в горячих местах программы в случаях, когда копии строк не нужны могут стать проблемой (снизится производительность сервера).

То же самое касается передачи `shared_ptr` -- копия этого объекта не бесплатная, задействуется внтуренний мютекс (или другой механзим защиты критической секции, защищающий счетчик ссылок). Но копии тут видимо нужна. Можно ее сделать в member initialization list (как у тебя, но сигнатуру констуктора надо изменить на `const&`), либо сделать `move` в member initialization list.

a-zagaevskiy commented on server/src/HTTP_Server.cpp:
```diff
@@ -7,12 +9,18 @@
 
-        auto message_handler = std::make_shared<MessageHandler>(ioc);
+        // Используется порт, указанный в параметрах
+        auto message_handler = std::make_shared<MessageHandler>(ioc, params.port);
```
На заметку: если создание объекта подразумевает выполнение определенной последовательности операций, то лучше инициализацию объета производить в специальный функциях фабриках. Это поможет избежать некорретной инициализации объекта, например, на стеке или от использования неверного типа умного указателя.
Пример:
```
auto message_handelr = std::make_unique<MessageHandler>(ioc, params.port);
message_handler->startHanlde();
``` 
такой код скомпилируется, однако, ошибка вылезет уже в runtime (во время выополенния программы), что очень неприятно.

a-zagaevskiy commented on server/src/Session.cpp:
```diff
@@ -47,46 +56,68 @@ void Session::sendNextPart(size_t offset, size_t total_size, std::string image_i
+            response->body() = "Server is busy. Try again later.";
+            response->prepare_payload();
+
+            http::async_write(socket, *response,
```
Кстати, недавно рассматривали этот момент с коллегами, использование `shared_ptr` для `response` нецелесообразно, т.к. boost::beast (boost::asio) тут все равно сделает копию.

Чтобы копии избежать, надо использовать другую перегрузку, кажется была с итераторами на буфер.
Но думаю, сейчас достаточно будет убарть аллокацию на куче (т.е. `response` создавать на стеке).

antekiv replied:
@a-zagaevskiy, а почему будет копия? 
По сигнатуре функция принимает ссылку:
https://www.boost.org/doc/libs/develop/libs/beast/doc/html/beast/ref/boost__beast__http__async_write/overload2.html

При этом пишется, что `msg` не владеет ресурсом:
```
msg | The message to write. The object must remain valid at least until the handler is called; ownership is not transferred.
```

В отличии, например от `std::thread`, где параметры явно передаются через универсальную ссылку, по которой они либо копируются либо муваются, и чтобы решить эту проблему, нужна обертка (`std::ref/std::cref`)

a-zagaevskiy reviewed (commented):
Илья, спасибо за работу! Посмотрел ревью, оставил несколько комметариев общего характера.

mayonezzzzzzz committed:
fix(client): optimization of creating directories

mayonezzzzzzz committed:
fix(client): comments removed

mayonezzzzzzz committed:
fix: creating variable for using in a "for" loop

mayonezzzzzzz committed:
fix(server): main thread reservation

mayonezzzzzzz committed:
fix(server): using prefix forms of increments and decrements

Avoid creating a temporary object to store current value of a variable

antekiv commented on client/src/HTTP_Client.cpp:
```diff
@@ -11,37 +11,19 @@ using tcp = asio::ip::tcp;
+bool checkDirectory(std::string& path) {
+    std::error_code ec;
+    if (!std::filesystem::is_directory(path, ec)) {
+        std::cerr << "Invalid directory - " << path << "\n";
```
Надо вывести ошибку, если есть в `ec`. 
Через `ec.message()`

Точно не помню, возможно ошибкой будет, если у пользователя нет прав на чтение директории. Т.е. директория есть, а доступа к ней нет.

mayonezzzzzzz committed:
fix: print error message

mayonezzzzzzz committed:
fix: comment removed

mayonezzzzzzz committed:
fix: using vector<unsigned char> everywhere

To ensure that one data type is used everywhere

mayonezzzzzzz committed:
fix(parametrization): using current directory as default paths

mayonezzzzzzz committed:
fix(.gitignore): removing local files for ignoring

All local files that exist only on my PC were added to the global .gitignore_global file.

mayonezzzzzzz committed:
fix(CmakeLists.txt): using target_ methods

Using target_link_libraries and target_include_directories.

Using BoostConfig.cmake file for building (path to this file is indicated in CMAKE_PREFIX_PATH variable).

mayonezzzzzzz committed:
fix(CmakeLists.txt): using find_package for OpenCV

Using find_package.

Specifying path to OpenCVConfig.cmake file in CMAKE_PREFIX_PATH variable.

mayonezzzzzzz committed:
fix(CmakeLists.txt): using find_package for Libjpeg-turbo

Using find_package.

Specifying path to libjpeg-turboConfig.cmake file in CMAKE_PREFIX_PATH variable.

antekiv commented on server/CMakeLists.txt:
```diff
@@ -10,10 +10,7 @@ set(SERVER_SOURCES
 add_executable(http_server ${SERVER_SOURCES})
 
 find_package(OpenCV REQUIRED)
+find_package(libjpeg-turbo REQUIRED CONFIG)
```
Я бы предложил эти `find_package` вынести на уровень выше, чтобы как можно раньше пользователь узнал, что нет какой-либо библиотеки

antekiv replied:
к тем, где ищется `boost`. Иначе cmake много лишних команд сделает, прежде, чем узнает, что какой-либо либы нету, и не получится собрать проект

mayonezzzzzzz replied:
Да, так правда будет лучше, сделаю так.

antekiv commented on client/src/HTTP_Client.cpp:
```diff
@@ -15,6 +15,7 @@ bool checkDirectory(std::string& path) {
     std::error_code ec;
     if (!std::filesystem::is_directory(path, ec)) {
         std::cerr << "Invalid directory - " << path << "\n";
+        std::cerr << "Error: " << ec.message() << "\n";
```
```cpp
bool checkDirectory(std::string& path) {
    std::error_code ec;
    bool isDirectory = std::filesystem::is_directory(path, ec);
    
    if (!isDirectory || ec) {
        std::cerr << "Invalid directory - " << path << "\n";
        std::cerr << "Error: " << ec.message() << "\n";
        path = ".";
        return false;
    }
    return true;
}
```
`ec` тоже нужно проверить, но этот код мне все равно не очень нравится...

antekiv replied:
как-будто не нужна такая обертка над `std::filesystem::is_directory`

antekiv replied:
тут не согласуется имя с тем, что происходит внутри - то что мне не нравится.
`checkDirectory` - подразумевает, что будет какая-то проверка, а тут еще устанавливается значение по умолчанию.

mayonezzzzzzz committed:
fix(CmakeLists.txt): all find_packeges are in main CmakeLists

mayonezzzzzzz committed:
feat: sending text as JSON metadata

Using Boost::json to send and process text for overlaying.

JSON data sends in the first request, image - in next requests.

mayonezzzzzzz committed:
feat(CmakeLists.txt): added Boost::json module

mayonezzzzzzz committed:
fix: size_t is using instead of std::atomic

mayonezzzzzzz committed:
fix: use "const &" for parametets in MessageHandler constructor

antekiv commented on server/src/Session.cpp:
```diff
@@ -73,58 +76,76 @@ void Session::handleRead(boost::system::error_code& err, std::size_t) {
+        auto& request_body = request.body();
+        std::vector<char> body_data(request_body.begin(), request_body.end());
+        std::cout << "Received part size: " << request_body.size() << std::endl;
+        request_body.clear();
```
Это необязательно делать, после выхода за скоуп вызовется деструктор у `request_body`, который сам все сделает

antekiv commented on server/src/Session.cpp:
```diff
@@ -11,8 +11,8 @@ namespace json = boost::json;
 const size_t BUFFER_SIZE = 1024 * 1024;
 
 // Текущее количество обрабатываемых запросов
+size_t MAX_REQUESTS = 0;
```
Комментарий к строке как-то не подходит..

mayonezzzzzzz replied:
Ой, поменял строки местами

mayonezzzzzzz committed:
fix: comment in the right place

mayonezzzzzzz committed:
fix(server): fabric method for creating MessageHandler object

mayonezzzzzzz committed:
fix(server): creating response on stack, not the heap

mayonezzzzzzz committed:
fix(client): checking error_code status

mayonezzzzzzz committed:
fix(client): function renamed to reflect its purpose

mayonezzzzzzz committed:
fix(server): request resets bofore next call of async_read

antekiv commented on server/src/MessageHandler.cpp:
```diff
@@ -9,6 +9,10 @@ MessageHandler::MessageHandler(const std::shared_ptr<asio::io_context>& ioc, con
 }
 
+std::shared_ptr<MessageHandler> MessageHandler::createMessageHandler(const std::shared_ptr<asio::io_context>& ioc, const std::string& port) {
+    return std::shared_ptr<MessageHandler>(new MessageHandler(ioc, port));
```
c C++14 появился `std::make_shared`, который так же создает умный указатель, но является более безопасным (нет явного вызова new, для которого в некоторых ситуациях не вызывается delete, если произошло исключение, или какая-то ошибка)

mayonezzzzzzz committed:
fix: make_shared instead of "new"

Using make_shared instead of allocating memory with new.

Avoiding memory leaks while exceptions happen.

mayonezzzzzzz committed:
fix(server): all objects are creating with fabric method

antekiv commented:
@mayonezzzzzzz, выглядит хорошо, много осталось исправить?

mayonezzzzzzz committed:
fix(server): fix in factory method

Object creation and method for start are moved tot a single factory method

a-zagaevskiy commented on server/src/Session.h:
```diff
@@ -15,6 +15,7 @@ extern size_t MAX_REQUESTS;
 class Session : public std::enable_shared_from_this<Session> {
 public:
 	Session(tcp::socket&& socket);
+	static std::shared_ptr<Session> createSession(tcp::socket&& socket);
```
Обычно еще играются с видимости констуктора и инициализирующих методов (их объявляют в private-секции).

mayonezzzzzzz replied:
Насчёт этого мы с Антоном говорили, в моем случае make_shared не может получить доступа к приватному конструктору, и возникает ошибка, поэтому здесь я оставил как есть

a-zagaevskiy replied:
Безопасность кода должна быть в приоритете. Если конструктор и инициализацию не спрятать, то потенциально можно допустить следующие ошибки:
* создать объект на стеке
* создать объект в куче, но не shared_ptr
* забыть позвать иницилизацию
* позвать инициализацию несколько раз

mayonezzzzzzz replied:
Хорошо, я буду иметь ввиду.
